New maintainer guide

oleg-nenashev opened the pull request:
Follow-up to https://groups.google.com/forum/#!topic/jenkinsci-dev/1ioTdU4784M

### Proposed changelog entries

If we want to highlight it...

* Introduce a new Jenkins core maintainer guide
  * https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc

oleg-nenashev committed:
Draft the new maintainer guide

res0nance commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -52,5 +52,5 @@ Before the changes are marked as `ready-for-merge`:
 - [ ] Proper changelog labels are set so that the changelog can be generated automatically
 - [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
-- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`
+- [ ] If it would make sense to backport the change to LTS, a Jira issue should exist and be labeled as `lts-candidate`
```
If its required to backport maybe the wording should be clear here.
```suggestion
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist and be labeled as `lts-candidate` to be considered
```

res0nance commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+There is no special preconditions to do so, anyone is welcome to contribute.
+
+**Core Pull Request Reviewers** is the next step for contributors who are interested to review Jenkins pull requests on a regular basis.
+They get [Triage permissions](TODO) so that they can manage pull requests, request reviews and prepare changelog drafts in the pull request description.
```
Perhaps this would make sense what exactly triage permissions is
```suggestion
They get [Triage permissions](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization) so that they can manage pull requests, request reviews and prepare changelog drafts in the pull request description.
```

res0nance commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+**Step 1: Maintainer checklist**. Merge process can be initiated once a pull request matches the requirements:
+
+* Pull request is compliant wuth requirements to submitters (see the link:/.github/PULL_REQUEST_TEMPLATE.md[pull request template])
```
```suggestion
* Pull request is compliant with requirements to submitters (see the link:/.github/PULL_REQUEST_TEMPLATE.md[pull request template])
```

res0nance commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set
+  and there is a `Proposed upgrade guidelines` section in the PR title.
+  (link:https://github.com/jenkinsci/jenkins/pull/4387[example])
+* It would make sense to backport the change to LTS, a Jira issue should exist and be labeled as `lts-candidate`
```
Same as above

res0nance replied:
@oleg-nenashev this should be as above stating that a jira should exist for backporting changes
```suggestion
* If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).
```

res0nance commented:
One more point. Does 2 approvals mean approvals from core-maintainers, core-pr-reviewers? Explicitly whose approval is required might be worth specifying

timja commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+= Documentation for core maintainers
+
+Jenkins core is a central component of the project which serves millions of users, and ot is critical to maintain the it in a good shape.
```
```suggestion
Jenkins core is a central component of the project which serves millions of users, and it is critical to maintain the it in a good shape.
```

daniel-beck replied:
"maintain the it"?

I don't know what the sentence should be, so no suggestion from me.

timja commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+**Core Pull Request Reviewers** is the next step for contributors who are interested to review Jenkins pull requests on a regular basis.
+They get [Triage permissions](TODO) so that they can manage pull requests, request reviews and prepare changelog drafts in the pull request description.
+GitHub team: link:TODO[@jenkinsci/core-pr-reviewers].
```
```suggestion
GitHub team: link:https://github.com/orgs/jenkinsci/teams/core-pr-reviewers[@jenkinsci/core-pr-reviewers].
```

timja commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+   and different files have different designs
+** Main goal is 
+* Make contributors to fix unrelated issues
+** Create follow-up issues instead, it is fine to referene them in comments so that the contributor might want to pick them up
```
```suggestion
** Create follow-up issues instead, it is fine to reference them in comments so that the contributor might want to pick them up
```

timja commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+* `ready-for-merge` label is set
+* An explicit comment is added to the pull request so that other repository watchers are notified.
+  Example: `Thanks to all contributors! We consider this change as read to be merged towards the next weekly release. It may be merged after a 24hours if there is no negative feedback`
```
```suggestion
  Example: `Thanks to all contributors! We consider this change as ready to be merged towards the next weekly release. It may be merged after 24hours if there is no negative feedback`
```

timja commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+//TODO(oleg_nenashev): Define "trivial" better to avoid loopholes
+** trivial changes like typo fixes or minor Javadoc improvements
+** changes which do not affect the production code: Jenkinsfile tweaks, tools inside the repo, etc. 
+
```
may not need to be stated, but possibly could add:
```suggestion
** broken master build

```

daniel-beck replied:
Agree. Cleaning up after bad merges should be expedited to unblock all other PRs and development.

res0nance replied:
I'm on board with this given that we were bitten by this exact scenario recently in #4467

timja commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+== Feedback
+
+The process documented in this document is not slate in stone.
```
```suggestion
The process documented in this document is not set in stone.
```

timja reviewed (changes requested):
Some copy-editing and one suggestion

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* Jenkins core
+* Jenkins modules
+* Libraries included into the Jenkins core
+* Core components like Winstone, Extras Executable WAR, etc.
```
```suggestion
* Core components like Winstone, Executable WAR, etc.
```

That's just a GitHub repo naming convention.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+* link:https://groups.google.com/d/forum/jenkinsci-dev[Jenkins Developer Mailing List]
+* Chat: N/A, to be created
+* Office hours: N/A, to be created
```
No need to mention this or chat.

oleg-nenashev replied:
I would prefer to do it so that we add contacts in the nearest future

timja replied:
It can be added when they are created?

oleg-nenashev replied:
```suggestion
* Office hours: N/A, to be created
```

oleg-nenashev replied:
```suggestion
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+They get [Triage permissions](TODO) so that they can manage pull requests, request reviews and prepare changelog drafts in the pull request description.
+GitHub team: link:TODO[@jenkinsci/core-pr-reviewers].
+
+<!--TODO: consider switching to Admin-->
```
👎

oleg-nenashev replied:
```suggestion
```

Agreed

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+=== Ladder
+
+* Anyone can participate as a **committer** or as a **reviewer**.
```
Except that the initial reviewers team suggestion required them to have experience with core? Doesn't seem to add up. Is this deliberate? Do we want to maintain a team with people who've never been active in this repo?

daniel-beck replied:
Right, got confused by the naming of this vs. the next one.

oleg-nenashev replied:
i will just replace the first role by "contributor", should help

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+** If there is no consensus about the feasibility and implementation,
+   code reviewers are expected to suggest proper channels to maintainers
+* Ensure that the pull request matches the code quality practices: test automation, Javadoc for newly introduced APIs, etc.
+** There is no goal to make the code perfect tho
```
API additions for use in plugins need to be carefully considered, however. When in doubt, `@Beta` or `@NoExternalUse` to make them effectively not _real_ API additions.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* Last but not least,
+  guide newcomer contributors who are not familiar with the project processes.
+
+=== Code review non-goals
```
It might not be clear what this means, so please explain.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* Enforce a particular coding style
+** Jenkins core has a complex codebase created by many contributors,
+   and different files have different designs
+** Main goal is 
```
Unfinished sentence.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+=== Code review non-goals
+
+* Enforce a particular coding style
```
Honestly, if you touch it and introduce a weird coding style, you should be called out for it.

oleg-nenashev replied:
Clarified it

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* Breakdown the commit history or squashing PRs
+** Not every contributor is a Git expert, do not request changes in the commit history unless it is necessary
+** Core maintainers can squash PRs during the merge.
+   If you feel this is important, use the `squash-me-please` label
```
```suggestion
   If you feel this is important, use the `squash-merge-me` label
```

daniel-beck replied:
When would it be important though?

Needs some guideline. Something like "An indicator to use merge commits would be if the PR history cleanly separates steps in developing the complete solution, that are individually useful to retain for the future. Substantial iteration after code review feedback, or a generally 'unclean' history (e.g. commits that fix compile errors, have no useful messages, or only exist to rebuild the PR) would indicate a preference for squash-merging the PR" perhaps?

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+=== Code review goals
+
+* Ensure that the pull request is feasible.
```
Don't unnecessarily write Powerpoint slides in text documents. This isn't for consumption by managers, we can handle actual paragraphs longer than one sentence.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+** Jenkins core has a complex codebase created by many contributors,
+   and different files have different designs
+** Main goal is 
+* Make contributors to fix unrelated issues
```
Needs clarification what this is about. Also tie back to keeping PRs focused.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+* Pull request is compliant wuth requirements to submitters (see the link:/.github/PULL_REQUEST_TEMPLATE.md[pull request template])
+* There are at least 2 approvals for the pull request and no outstanding requests for change
+* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
```
```suggestion
* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change (often indicated by line comments attached to an approving PR review, or by using the term "nit", from "nit-picking")
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* Pull request is compliant wuth requirements to submitters (see the link:/.github/PULL_REQUEST_TEMPLATE.md[pull request template])
+* There are at least 2 approvals for the pull request and no outstanding requests for change
+* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
+* Changelog entries in the PR title and/or `Proposed changelog entries` are correct
```
```suggestion
* Changelog entries in the PR title and/or `Proposed changelog entries` are correct and reflect the current, final state of the PR
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* There are at least 2 approvals for the pull request and no outstanding requests for change
+* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
+* Changelog entries in the PR title and/or `Proposed changelog entries` are correct
+* Proper changelog labels are set so that the changelog can be generated automatically
```
Which labels are changelog-relevant?

timja replied:
this should be set on the label description imo.

daniel-beck replied:
We need labels for labels!

oleg-nenashev replied:
I will just reference https://github.com/jenkinsci/.github/blob/master/.github/release-drafter.yml for now to K.I.S.S

daniel-beck replied:
Oooooh that a nice solution 👍 @oleg-nenashev

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
+* Changelog entries in the PR title and/or `Proposed changelog entries` are correct
+* Proper changelog labels are set so that the changelog can be generated automatically
+* If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set
```
I don't think this correctly describes what it takes to get an upgrade guide entry.

How about:

```suggestion
* If the change needs administrators to be aware of it when upgrading, the `upgrade-guide-needed` label is present. This is usually the case when a data migration occurs, a feature has been removed, a significant behavior change is introduced (including when there is a way to opt out), or in general when we expect at least a large minority of admins to benefit from knowing about the change, e.g. to apply a new option.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+  Common review process is used otherwise.
+* 24 hours timeout after `ready-for-merge` is not required for...
+//TODO(oleg_nenashev): Define "trivial" better to avoid loopholes
+** trivial changes like typo fixes or minor Javadoc improvements
```
```suggestion
** changes that do not result in changes to the primary deliverables, such as minor changes to code commits, or typo fixes in local (non-parameter) variables.
```

oleg-nenashev replied:
I need to think about it. there was no intention to focus on variables only, typo fixes in WebUI also qualify IMHO

oleg-nenashev replied:
```suggestion
* changes that do not result in changes to the primary functionality, such as typo fixes in documentation or code which is not exposed as public API.
```

daniel-beck replied:
@oleg-nenashev Disagree, that would mean any `internal` PR (e.g. big refactorings and rewrites) is fair game to merge quickly.

oleg-nenashev replied:
```suggestion
** changes that do not result in changes to the primary functionality, such as typo fixes in documentation or localization files
```

What about that @daniel-beck ? If not, will drop the bullet entirely

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+
+=== Squashing pull requests OR not?
+
+Right now we do not have a strict policy about it.
```
Would reference my suggestion from above.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,162 @@
+but also that they do not introduce undesired defects, breaking changes and technical debt.
+At the same time, we are interested to make the review process as simple as possible for contributors and maintainers.
+
+=== Code review goals
```
This is missing that reviewers are expected to look at changes critically from a "product management" point of view. It's not _just_ about the code, but also whether the change makes sense in a global / holistic way, considering existing popular plugins and the way users experience Jenkins overall.

daniel-beck replied:
Also: Make it explicit that it is OK to reject changes for pretty much any reason based in the change itself, or request that they (especially in the case of complex changes) go through the JEP process first. Following these guidelines, we require consensus but do not "allow" reviewers to object to a PR for any non-technical reason.

oleg-nenashev replied:
I mostly addressed it, except the `Following these guidelines, we require consensus but do not "allow" reviewers to object to a PR for any non-technical reason.` phrase

daniel-beck replied:
To clarify, that wasn't prescriptive, but rather describing a problem I saw with the guidelines. The new/updated section captures my goal very nicely.

oleg-nenashev committed:
Maintainer guide - Apply suggestions from code review

Co-Authored-By: Raihaan Shouhell <raihaanhimself@gmail.com>
Co-Authored-By: Tim Jacomb <t.jacomb@kainos.com>
Co-Authored-By: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

oleg-nenashev commented:
Thanks all for feedback, and sorry for the mess in the initial PR (`olegs-plane-project`). Will address all comments

jeffret-b reviewed (commented):
I really like the improvement to the process and the document as presented here. This is going to help keep things flowing.

My biggest concern is with the naming of the different roles. As I describe in my comments, it's hard to keep them separate and the names aren't clearly distinguishable. I'm not sure what we need to do to fix it.

I left several minor comments about typos, wording, or grammar.

It might be helpful to include some comments about expected conduct. The items that come to mind include thanking contributors for their work, reviewers being responsive to follow-on and not blocking PRs by simply not responding, etc. Maybe there should be a time limit for when a committer has responded and tried to get updated feedback from a reviewer but the reviewer is unresponsive. Maybe also something about encouraging reviewers to provide positive, useful feedback: Instead of just saying "this is wrong", point committers to correct examples or suggest how it could work correctly.

jeffret-b commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,163 @@
+=== Maintainer roles
+
+* Committer / reviewer
+* Core Pull Request Reviewer
```
I think part of the trouble I'm having with suggesting better naming is that I'm not yet entirely clear on the different roles. The level 1 role (basic GH user) and the level 4 role (release manager) are pretty clear. I'm not finding the level 2 and 3 roles as clear so it's harder to suggest meaningfully distinctive names.

fcojfernandez commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,163 @@
+* Anyone can participate as a **committer** or as a **reviewer**.
+  There is no precondition except having a GitHub account, just submit pull requests or comment in existing ones!
+* **Core Pull Request Reviewers** can be nominated by contributors in the link:https://groups.google.com/d/forum/jenkinsci-dev[developer mailing list]. 
+Self-nomination are fine.
```
No native speaker, but it sounds odd to me
```suggestion
Self-nomination is fine.
```

fcojfernandez reviewed (commented):
I really like this documentation work and I totally agree with @jeffret-b 's general comment.

Thanks @oleg-nenashev !

oleg-nenashev commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,163 @@
+
+**Step 3: Merge!**.
+A Core Maintainer merges the change after a timeout (if needed).
+After that the change will be landed in the next weekly release unless it gets reverted later.
```
```suggestion
After that the change will be landed in the next weekly release.
```

oleg-nenashev committed:
Apply suggestions from code review

Co-Authored-By: Jeff Thompson <37345299+jeffret-b@users.noreply.github.com>
Co-Authored-By: Francisco Fernández <31063239+fcojfernandez@users.noreply.github.com>

oleg-nenashev committed:
Maintainer guide - Replace the first role by "contributor"

oleg-nenashev committed:
Maintainer guide - Add more details

oleg-nenashev committed:
Maintainer guide - address some feedback

jeffret-b commented:
This is coming along nicely. I like the change to calling the first level, "Contributors". The different levels are now reasonably clear and distinct. I've left a few minor comments.

oleg-nenashev commented:
@jeffret-b Thanks for the feedback! W.r.t "I've left a few minor comments.", I suspect you have not submitted your review yet

> This is coming along nicely. I like the change to calling the first level, "Contributors". The different levels are now reasonably clear and distinct. I've left a few minor comments.

fcojfernandez commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,163 @@
+** Main goal is 
+* Make contributors to fix unrelated issues
+** Create follow-up issues instead, it is fine to reference them in comments so that the contributor might want to pick them up
+* Breakdown the commit history or squashing PRs
```
Agree with Jeff

daniel-beck replied:
Needs clarification.

oleg-nenashev replied:
```suggestion
* Make contributors to have atomic commit history or to squash their pull request
```

timja reviewed (approved):
Added a couple of comments and there's a few other minor comments I've 👍'ed,

nothing blocking that I can see though and this is far better than the current state (nothing)

jeffret-b commented:
> I suspect you have not submitted your review yet

Yes, I clicked the wrong button. I've just submitted that.

jeffret-b reviewed (approved):
I think this could benefit from a few more tweaks as several people have commented to polish it off a bit more. On the other hand it's certainly better than what we currently have so I'll approve it. I consider the comments nice-to-have and not blocking.

daniel-beck commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -27,7 +27,7 @@ N/A
 ### Submitter checklist
 
-- [ ] JIRA issue is well described
+- [ ] Jira issue is well described (not mandatory)
```
Ambiguous:

* Is it not necessary to have a Jira issue?
* Is it not necessary that the Jira issue is useful?

oleg-nenashev replied:
What if I adjust it like...

```suggestion
- [ ] Jira issue is well described (if needed)
```

Can just revert it tho

daniel-beck replied:
How about

```suggestion
- [ ] (If applicable) Jira issue is well described
```

daniel-beck commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -52,5 +52,4 @@ Before the changes are marked as `ready-for-merge`:
 - [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
-- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`
-
+- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist and be labeled as `lts-candidate` to be considered
```
```suggestion
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).
```

daniel-beck commented on CONTRIBUTING.md:
```diff
@@ -107,8 +107,9 @@ It is a good practice is to create branches instead of pushing to master.
-  * If you do not get feedback after 3 days, feel free to ping `@jenkinsci/code-reviewers` to CC.
-  * Usually we merge pull requests after 2 votes from reviewers or after 1 vote and 1-week delay without extra reviews.
+  * If you do not get feedback after 3 days, feel free to ping `@jenkinsci/core-pr-reviewers` to CC.
+  * Usually we merge pull requests after 2 votes from reviewers and some feedback timeout.
```
```suggestion
  * Usually we merge pull requests after 2 approvals from reviewers, no requested changes, and having waited some more time to give others an opportunity to provide more feedback.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+There is no special preconditions to do so, anyone is welcome to contribute.
+
+**Core Pull Request Reviewers** is the next step for contributors who are interested to review Jenkins pull requests on a regular basis.
+They get [Triage permissions](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization) so that they can manage pull requests, request reviews and prepare changelog drafts in the pull request description.
```
```suggestion
They get https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization[Triage permissions] so that they can manage pull requests, request reviews and prepare changelog drafts in the pull request description.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+
+**Release Team Members** are responsible for Jenkins weekly and LTS releases.
+Led by the link:https://jenkins.io/project/team-leads/#release[Jenkins Release Officer], they initiate releases, prepare changelogs and backport changes into the link:TODO[Stable branch].
+Team members get `Admin` permissions in the Jenkins core repository, and they also get permissions to trigger release Pipelines.
```
```suggestion
Team members get `Write` permissions in the Jenkins core repository, and they also get permissions to trigger release Pipelines.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+GitHub team: link:https://github.com/orgs/jenkinsci/teams/core[@jenkinsci/core].
+
+**Release Team Members** are responsible for Jenkins weekly and LTS releases.
+Led by the link:https://jenkins.io/project/team-leads/#release[Jenkins Release Officer], they initiate releases, prepare changelogs and backport changes into the link:TODO[Stable branch].
```
Makes no sense to insert a link here when the stable branch changes every few months.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+They are also responsible to monitor the weekly release status and to perform triage of critical issues.
+GitHub team: link:https://github.com/orgs/jenkinsci/teams/core[@jenkinsci/core].
+
+**Release Team Members** are responsible for Jenkins weekly and LTS releases.
```
Would comment this section out while it's still pretty much a WIP.

oleg-nenashev replied:
I agree that we should rater keep it

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+
+=== Ladder
+
+* *Contributors*. Anyone can participate.
```
```suggestion
* **Contributors**. Anyone can participate.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+
+There are the following code review goals:
+
+* Ensure that the pull request is feasible and that it fits the Jenkins ecosystem
```
```suggestion
* Ensure that the pull request is _feasible_ and that it fits the Jenkins ecosystem
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+There are the following code review goals:
+
+* Ensure that the pull request is feasible and that it fits the Jenkins ecosystem
+* Verify compatibility of the suggested changes
```
```suggestion
* Verify _compatibility_ of the suggested changes
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+
+* Ensure that the pull request is feasible and that it fits the Jenkins ecosystem
+* Verify compatibility of the suggested changes
+* Ensure that the pull request matches the code quality practices: test automation, Javadoc for newly introduced APIs, etc.
```
```suggestion
* Ensure that the pull request matches the _code quality_ practices: test automation, Javadoc for newly introduced APIs, etc.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Verify compatibility of the suggested changes
+* Ensure that the pull request matches the code quality practices: test automation, Javadoc for newly introduced APIs, etc.
+* Ensure that the pull request contains appropriate changelogs and, if needed, upgrade guidelines.
+* Ensure there is a consensus among reviewers about all items above 
```
```suggestion
* Ensure there is a _consensus_ among reviewers about all items above.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Ensure that the pull request is feasible and that it fits the Jenkins ecosystem
+* Verify compatibility of the suggested changes
+* Ensure that the pull request matches the code quality practices: test automation, Javadoc for newly introduced APIs, etc.
+* Ensure that the pull request contains appropriate changelogs and, if needed, upgrade guidelines.
```
This doesn't have its own section below, as the only item in this list. Deliberate?

oleg-nenashev replied:
yes. Can be expanded later if needed

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+We accept breaking changes when it is needed (security fixes, feature deprecation and removal, etc.),
+but ultimately the project's goal is retain as much compatibility as possible.
+It includes both feature compatibility and binary/API compatibility which is important for the plugin ecosystem.
+Although we provide some tools for checking API usages across open-source plugins,
```
```suggestion
Although we have some tools (like https://github.com/jenkins-infra/usage-in-plugins[usage-in-plugins]) for checking API usages across open-source plugins,
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+but ultimately the project's goal is retain as much compatibility as possible.
+It includes both feature compatibility and binary/API compatibility which is important for the plugin ecosystem.
+Although we provide some tools for checking API usages across open-source plugins,
+there is no way to confirm external usages in 3rd-party proprietary plugins which is also a part of the ecosystem.
```
Would not consider that relevant. It's also not actionable, what would a reader take from this?

oleg-nenashev replied:
![image](https://user-images.githubusercontent.com/3000480/74657443-8c58ed00-5190-11ea-9466-bb2f5305ce29.png)

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+test coverage for newly added functionality and fixes,
+documentation for newly introduced APIs,
+the submitted code is readable and matches the code style in the surrounding codebase,
+etc.
```
Add a note that we're aware that there are existing inconsistencies in the code, and that new code should follow the (majority) style guide, while updates to existing code should only fix formatting on the lines affected anyway to keep the diff minimal, while respecting the overall formatting of the file?

MarkEWaite replied:
I also would appreciate a recommendation to keep diff volume small.  

Minimizing diff size helps reviewers focus their attention on the specifics of the change and reduces the risk of a change from one pull request creating a conflict in another pull request.  Both are worthwhile.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+Pull requests often become a venue to discuss feasibility, underlying technical decisions and design.
+We are fine with that as well.
+If there is no consensus about the feasibility and implementation,
+code reviewers are expected to suggest proper channels to maintainers.
```
```suggestion
code reviewers are expected to suggest proper channels for contributors to discuss their contribution.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+If there is no consensus about the feasibility and implementation,
+code reviewers are expected to suggest proper channels to maintainers.
+
+* A discussion in the link:https://groups.google.com/d/forum/jenkinsci-dev[Jenkins Developer Mailing List] is a default way to go
```
```suggestion
* A discussion in the link:https://groups.google.com/d/forum/jenkinsci-dev[Jenkins Developer Mailing List] is the default way to go
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+code reviewers are expected to suggest proper channels to maintainers.
+
+* A discussion in the link:https://groups.google.com/d/forum/jenkinsci-dev[Jenkins Developer Mailing List] is a default way to go
+* If a consensus cannot be reached in the mailing list,
```
```suggestion
* If no consensus can be reached on the mailing list,
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+
+* A discussion in the link:https://groups.google.com/d/forum/jenkinsci-dev[Jenkins Developer Mailing List] is a default way to go
+* If a consensus cannot be reached in the mailing list,
+  voting at the link:https://jenkins.io/project/governance-meeting/[Jenkins Governance Meeting] can be used to get a final decision
```
```suggestion
  voting at the link:https://jenkins.io/project/governance-meeting/[Jenkins Governance Meeting] can be used to get a final decision.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Breakdown the commit history or squashing PRs
+** Not every contributor is a Git expert, do not request changes in the commit history unless it is necessary
+** Core maintainers can squash PRs during the merge.
+   If you feel this is important, use the `squash-merge-me` label
```
```suggestion
   If you feel this is important, add the `squash-merge-me` label
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+** Not every contributor is a Git expert, do not request changes in the commit history unless it is necessary
+** Core maintainers can squash PRs during the merge.
+   If you feel this is important, use the `squash-merge-me` label
+** We want to keep pull requests atomic when possible (one feature / fix - separate commit),
```
```suggestion
** We want to keep pull requests focused when possible (one feature / fix per pull request),
```

daniel-beck replied:
What does "separate commit" mean here?

oleg-nenashev replied:
Whatever. Your option is better

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change (often indicated by line comments attached to an approving PR review, or by using the term "nit", from "nit-picking")
+* Changelog entries in the PR title and/or `Proposed changelog entries` are correct and reflect the current, final state of the PR
+* Proper changelog labels are set so that the changelog can be generated automatically
+* If the change needs administrators to be aware of it when upgrading, the `upgrade-guide-needed` label is present. This is usually the case when a data migration occurs, a feature has been removed, a significant behavior change is introduced (including when there is a way to opt out), or in general when we expect at least a large minority of admins to benefit from knowing about the change, e.g.  ...
```
Line could use some reformatting after applying the suggested edit.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Changelog entries in the PR title and/or `Proposed changelog entries` are correct and reflect the current, final state of the PR
+* Proper changelog labels are set so that the changelog can be generated automatically
+* If the change needs administrators to be aware of it when upgrading, the `upgrade-guide-needed` label is present. This is usually the case when a data migration occurs, a feature has been removed, a significant behavior change is introduced (including when there is a way to opt out), or in general when we expect at least a large minority of admins to benefit from knowing about the change, e.g.  ...
+  and there is a `Proposed upgrade guidelines` section in the PR title.
```
Broken text.

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+Decision is made by a consensus in the mailing list or via voting at the link:https://jenkins.io/project/governance-meeting/[governance meeting].
+* **Core Maintainers** - same process as for Core PR reviewers.
+  All nominees must also sign an link:https://github.com/jenkinsci/infra-cla/[Individual Contributor License Agreement] before getting the permission in GitHub repositories.
+* **Release Team Members** are assigned by the Jenkins Release officer
```
Strictly speaking, the final boss is the release officer, elected every year? 😄

oleg-nenashev replied:
Yup

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+there is no way to confirm external usages in 3rd-party proprietary plugins which is also a part of the ecosystem.
+
+**Code quality**.
+There is no goal to make the code perfect, but we want to ensure that all new code matches basic quality standards:
```
Weird phrasing, perhaps

```suggestion
The code doesn't have to be perfect, but we want to ensure that all new code matches basic quality standards:
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+  Security hardening and enhancements go through the standard process.
+* Release Team members are permitted to bypass the review/merge process if and only if a change is needed to unblock the security release.
+  Common review process is used otherwise.
+* 24 hours timeout after `ready-for-merge` is not required for...
```
```suggestion
* 24 hours waiting perioid after adding the `ready-for-merge` label is not required for:
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* An explicit comment is added to the pull request so that other repository watchers are notified.
+  Example: `Thanks to all contributors! We consider this change as ready to be merged towards the next weekly release. It may be merged after 24hours if there is no negative feedback`
+
+**Step 3: Merge!**.
```
```suggestion
**Step 3: Merge!**
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+
+* `ready-for-merge` label is set
+* An explicit comment is added to the pull request so that other repository watchers are notified.
+  Example: `Thanks to all contributors! We consider this change as ready to be merged towards the next weekly release. It may be merged after 24hours if there is no negative feedback`
```
```suggestion
  Example: _Thanks to all contributors! We consider this change as ready to be merged towards the next weekly release. It may be merged after 24hours if there is no negative feedback_
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Pull request is compliant with requirements to submitters (see the link:/.github/PULL_REQUEST_TEMPLATE.md[pull request template])
+* There are at least 2 approvals for the pull request and no outstanding requests for change
+* Conversations in the pull request are over OR it is explicit that a reviewer does not block the change (often indicated by line comments attached to an approving PR review, or by using the term "nit", from "nit-picking")
+* Changelog entries in the PR title and/or `Proposed changelog entries` are correct and reflect the current, final state of the PR
```
```suggestion
* Changelog entries in the PR title and/or _Proposed changelog entries_ are correct and reflect the current, final state of the PR
```

oleg-nenashev committed:
Apply suggestions from code review by @daniel-beck

Co-Authored-By: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

oleg-nenashev committed:
Update .github/PULL_REQUEST_TEMPLATE.md

Co-Authored-By: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

oleg-nenashev committed:
Maintainer guide - more updates

oleg-nenashev commented:
I will try to get it over the line tomorrow

daniel-beck commented:
This should also explain merge timing considerations, which requires awareness of the [LTS schedule](https://jenkins.io/download/lts/) ([calendar](https://jenkins.io/event-calendar/)):

* Jenkins security updates are coordinated with the LTS calendar, and if the weekly introduces regressions, users of that line may have to choose between security fixes, and a working Jenkins.
* LTS baseline selection is also coordinated with the LTS calendar and happens in the week of the .3 release.

For these reasons, the following changes should not be merged during the week before LTS releases (weeks 3, 7, 11, 15, etc. on the page linked above):

* Changes that could be considered risky (relatively high risk of introducing regressions), as they could make users of Jenkins weekly releases choose between getting security fixes, and having a functioning Jenkins
* Very large changes (in terms of lines changed), because the Jenkins security team needs to prepare security fixes for the weekly release line in a very short period of time.

The Jenkins security team will usually send a "pre-announcement" to [the advisories list](https://groups.google.com/forum/#!forum/jenkinsci-advisories) on Wednesday or Thursday the week before release, but that's not always the case.

oleg-nenashev committed:
Maintainer Guide: Add info about the Security Releases and merges + add an explicit Jenkins LTS section

Addresses comments from @daniel-beck and @olivergondza

oleg-nenashev committed:
Maintainer guide: reference the advisories

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -161,18 +163,33 @@ Code reviews do NOT pursue the following goals:
 After that the change will be landed in the next weekly release.
 LTS Backporting, if needed, will be handled separately by the release team.
 
+Keep in mind that we set 
```
Unfinished

oleg-nenashev replied:
I was about linking the LTS section below, will fix

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -190,10 +207,18 @@ LTS Backporting, if needed, will be handled separately by the release team.
+
+Jenkins also offers the link:https://jenkins.io/download/lts/[LTS Release Line].
+It is maintained by the Jenkins Release Team which coordinates link:https://jenkins.io/download/lts/#backporting-process[backporting] and release candidate testing.
+It mostly happens through the developer mailing list and through backporting pull requests labeled with link:https://github.com/jenkinsci/jenkins/labels/into-lts[into-lts].
```
Ambiguous or wrong: These PRs go directly into LTS, they _are_ the backports.

oleg-nenashev replied:
```suggestion
It mostly happens through the developer mailing list. Backports are submitted as pull requests labeled with link:https://github.com/jenkinsci/jenkins/labels/into-lts[into-lts].
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -161,18 +163,33 @@ Code reviews do NOT pursue the following goals:
-a Core PR Reviewer or a Maintainer can mark the pull request for merge.
+==== Step 2: Is it a good time to merge?
+
+link:[https://jenkins.io/security/](Jenkins security updates) are coordinated with the LTS calendar, and if the weekly introduces regressions, users of that line may have to choose between security fixes, and a working Jenkins.
```
```suggestion
link:[https://jenkins.io/security/](Jenkins security updates) are coordinated with the LTS calendar, and if the weekly release on the weekend before an LTS release introduces regressions, users of the weekly line may have to choose between security fixes, and a working Jenkins.
```

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -161,18 +163,33 @@ Code reviews do NOT pursue the following goals:
+* Very large changes (in terms of lines changed), because the Jenkins security team needs to prepare security fixes for the weekly release line in a very short period of time.
+
+If the change is ready but it is not a good time, consider labeling the pull request with the `on-hold` label.
+If a merge is needed, please consult with the Jenkins Security officer
```
No such thing?

oleg-nenashev replied:
Hm? Do you suggest removing it?

daniel-beck replied:
I'm questioning the premise of "a merge is needed".

Should have submitted your PR earlier if you need something very large to go into a specific weekly release.

oleg-nenashev replied:
I will just remove it for now. There are always ways to escalate of merge is needed

daniel-beck commented on docs/MAINTAINERS.adoc:
```diff
@@ -161,18 +163,33 @@ Code reviews do NOT pursue the following goals:
+* Changes that could be considered risky (relatively high risk of introducing regressions), as they could make users of Jenkins weekly releases choose between getting security fixes, and having a functioning Jenkins
+* Very large changes (in terms of lines changed), because the Jenkins security team needs to prepare security fixes for the weekly release line in a very short period of time.
+
+If the change is ready but it is not a good time, consider labeling the pull request with the `on-hold` label.
```
```suggestion
If the change is ready but it is not a good time, consider labeling the pull request with the `on-hold` label.
Make sure to add a comment explaining why it was put on hold.
```

MarkEWaite commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,230 @@
+but ultimately the project's goal is retain as much compatibility as possible.
+It includes both feature compatibility and binary/API compatibility which is important for the plugin ecosystem.
+Although we have some tools (like https://github.com/jenkins-infra/usage-in-plugins[usage-in-plugins]) for checking API usages across open-source plugins,
+there is no way to confirm external usages in 3rd-party proprietary plugins which is also a part of the ecosystem.
```
Minor phrasing suuggestion
```suggestion
there is no way to confirm external usages in 3rd-party proprietary plugins which are also a part of the ecosystem.
```

MarkEWaite commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,230 @@
+* Enforcing a particular coding style.
+  Jenkins core has a complex codebase created by many contributors, and different files have different designs.
+  Our main goal is to firstly have the code readable by other contributors.
+* Make contributors to fix unrelated issues
```
Minor phrasing suggestion
```suggestion
* Make contributors fix issues that are not related to the primary topic of the pull request
```

MarkEWaite commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,230 @@
+* Very large changes (in terms of lines changed), because the Jenkins security team needs to prepare security fixes for the weekly release line in a very short period of time.
+
+If the change is ready but it is not a good time, consider labeling the pull request with the `on-hold` label.
+If a merge is needed, please consult with the Jenkins Security officer
```
```suggestion
If a merge is needed, please consult with the Jenkins Security officer.
```

MarkEWaite commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,230 @@
+After that the change will be landed in the next weekly release.
+LTS Backporting, if needed, will be handled separately by the release team.
+
+Keep in mind that we set 
```
Either an incomplete sentence or the start of the next section heading `Exceptions`.  Assumed an incomplete sentence.
```suggestion
```

MarkEWaite reviewed (approved):
One area of discomfort around:

> trivial changes like typo fixes or minor Javadoc improvements

that is well described in comments from @daniel-beck , but not enough to block the merge or prevent this from being a great improvement over what we have published.

I included a few optional suggestions for consideration.  All of them are optional.

oleg-nenashev committed:
Apply suggestions from code review

Co-Authored-By: Mark Waite <mark.earl.waite@gmail.com>
Co-Authored-By: Raihaan Shouhell <raihaanhimself@gmail.com>
Co-Authored-By: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

oleg-nenashev commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+* Release Team Member
+
+**Contributors** submit pull requests to the Jenkins core and review changes submitted by others.
+They can also review and triage incoming issues in Jenkins Jira.
```
I will add Jira separately once the triage process is established

```suggestion
```

oleg-nenashev commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,200 @@
+Led by the link:https://jenkins.io/project/team-leads/#release[Jenkins Release Officer], they initiate releases, prepare changelogs and backport changes into the link:TODO[Stable branch].
+Team members get `Admin` permissions in the Jenkins core repository, and they also get permissions to trigger release Pipelines.
+
+=== Ladder
```
I tried merging it, but such approach looks better. Roles are more important to standalone contributors

oleg-nenashev commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,231 @@
+* Breakdown the commit history or squashing PRs
+** Not every contributor is a Git expert, do not request changes in the commit history unless it is necessary
+** Core maintainers can squash PRs during the merge.
+   If you feel this is important, add the `squash-merge-me` label
```
```suggestion
   If you feel this is important, add the [squash-merge-me](https://github.com/jenkinsci/jenkins/pulls?q=is%3Aopen+is%3Apr+label%3Asquash-merge-me) label
```

oleg-nenashev committed:
Apply suggestions from code review

Co-Authored-By: Mark Waite <mark.earl.waite@gmail.com>

oleg-nenashev committed:
Fix typos

oleg-nenashev committed:
Merge branch 'maintainer-guide' of https://github.com/oleg-nenashev/jenkins into maintainer-guide

# Conflicts:
#	docs/MAINTAINERS.adoc

oleg-nenashev committed:
Expand code review goals: documentation and code styles

oleg-nenashev committed:
Maintainer guide: Document the squash-merge policy

oleg-nenashev commented:
I hope I addressed all comments. Major changes:

* Goals and non-goals are defined better. Documentation and changelogs expectations are now explicit
* There is squash-merge recommendations added as @daniel-beck suggested

oleg-nenashev commented:
I believe this guide is ready for a final review. All comments are addressed IMHO

MarkEWaite commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,272 @@
+
+==== Building consensus
+
+Not all changes are discussed before they submitted as pull requests.
```
```suggestion
Not all changes are discussed before they are submitted as pull requests.
```

MarkEWaite commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,272 @@
+
+When do we squash commits?
+
+* We squash commits during when core maintainers decide to do so (`squash-merge-me` label), usually when the conditions above are not met.
```
```suggestion
* We squash commits when core maintainers decide to do so (`squash-merge-me` label), usually when the conditions above are not met.
```

MarkEWaite reviewed (approved):
Two non-blocking minor phrasing comments.  Approved with sincere thanks for creating it!

olivergondza commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,272 @@
+but ultimately the project's goal is retain as much compatibility as possible.
+It includes both feature compatibility and binary/API compatibility which is important for the plugin ecosystem.
+Although we have some tools (like https://github.com/jenkins-infra/usage-in-plugins[usage-in-plugins]) for checking API usages across open-source plugins,
+there is no way to confirm external usages in 3rd-party proprietary plugins which are also a part of the ecosystem.
```
I do not want to rathole on this, but do we consider 3rd-party plugin to be part of the ecosystem? To me, it would make more sense to define the boundaries of the ecosystem (pretty much https://github.com/{jenkinsci,jenkins-infra,stapler}) and that would be the place where it is possible to make certain effort to ensure compatibility, but there is not commitment elsewhere.

oleg-nenashev replied:
I consider 3rd-party plugins as a part of wider Jenkins ecosystem which we would like to respect and minimize breaking changes there. Not just because I work for CloudBees, there we can easily fix breaking changes. But there are thousands of in-house plugins created by Jenkins users. I have written at least one at all my previous companies

olivergondza reviewed (approved):
This makes sense to me in general. Thanks!

jeffret-b reviewed (approved):
I read through it again. Looks good!

oleg-nenashev commented on docs/MAINTAINERS.adoc:
```diff
@@ -0,0 +1,272 @@
+=== Communication channels
+
+* link:https://groups.google.com/d/forum/jenkinsci-dev[Jenkins Developer Mailing List]
+* Chat: N/A, to be created
```
```suggestion
```

oleg-nenashev committed:
Apply suggestions from code review

Co-Authored-By: Mark Waite <mark.earl.waite@gmail.com>

oleg-nenashev commented:
@res0nance @daniel-beck I would like to get it merged if you are not against the change. We can always update it in follow-up pull requests

daniel-beck commented:
Would appreciate a response to https://github.com/jenkinsci/jenkins/pull/4472/files#r379001153

oleg-nenashev commented:
> Would appreciate a response to https://github.com/jenkinsci/jenkins/pull/4472/files#r379001153

Sorry, I did not resolve the comment. There is a special section added for code style

oleg-nenashev commented:
I plan to merge it tomorrow if no negative feedback

oleg-nenashev commented:
Sorry, I merged it slightly earlier than 24 hours. Should not be a problem since it does not impact the production code

Wadeck reviewed (approved):
:bee: Nice work Oleg!
